OPC UA Gateway thin-edge.io integration (#945)
SoftwareAG/c8y-docs: content/protocol-integration/opcua-bundle/gateway-register.md

Windows OS is used for the example.
{{< /c8y-admon-info >}}

> ### Thin-Edge

Reviewer: ```suggestion
### Thin Edge
```

> OPC UA Gateway can also be registered and operated via [thin-edge.io](https://thin-edge.io/). In contrast to standalone mode `thinEdge` configurations must be added to YAML file:

Author: added

> The OPC UA Gateway can also be registered and operated via [thin-edge.io](https://thin-edge.io/). In contrast to standalone mode `thinEdge` configurations must be added to YAML file:

Reviewer: ```suggestion
The OPC UA gateway can also be registered and operated via [thin-edge.io](https://thin-edge.io/). In contrast to the standalone mode, `thinEdge` configurations must be added to the YAML file:
```

which yaml file? *application-myTenant.yaml* from above? or another file?

Author: changed to you suggestion

```yaml
C8Y:
    baseUrl: https://<<yourTenant>>.{{< domain-c8y >}}
gateway:
    bootstrap:
        tenantId: <<yourTenantId>>
    identifier: Gateway_Device

>     name: My Gateway

Reviewer: not saying it's wrong, but since there's no quotes here - the name contains a space character, and like this, the capitalization on "Gateway" (as the second word) goes against our capitalization rules.

Author: changed to Gateway_Device in both identitfier and name

    db:
# The gateway uses the local database to store platform credentials and local cache. This parameter shows the location in which the local data should be stored.
        baseDir: C:/Users/<<userName>>/.opcua/data
    thinEdge:
        enabled: true
        mqttServerURL: tcp://<<thinEdge MQTT broker>>

>         deviceId: My Gateway

Reviewer: see above (name)

Author: see above

>         deviceId: Gateway_Device

Reviewer 2: shouldn't this value be the thinEdge deviceId?

```

> With the configuration `gateway.thinEdge.enabled: true` you switch to the thinEdge mode. Which means the authentication and registration to the platform will be done via thin-edge. The OPC UA gateway will automatically registered and created as sub-device under thin-edge device. The `gateway.thinEdge.mqttServerURL` and `gateway.thinEdge.deviceId` are the connection informations for the mqtt client to connect to the local thin-edge MQTT broker. 

Reviewer: ```suggestion
With the configuration `gateway.thinEdge.enabled: true` you switch to the thinEdge mode. This means that the authentication and registration to the platform will be done via Thin Edge. The OPC UA gateway is automatically registered and created as a sub-device under the Thin Edge device. `gateway.thinEdge.mqttServerURL` and `gateway.thinEdge.deviceId` are the connection information for the MQTT client to connect to the local Thin Edge MQTT broker. 
```

Author: changed to your suggestions

      # automatically when the legacyCleanup is set to true, which is the default setting.
      # If the legacy files wanted to be kept or if the mechanism for clearing is not needed, set legacyCleanup to false.
      legacyCleanup: true

>   # 

Reviewer 3: Why this empty comment?

Author: added comment

  thinEdge:
    # Enables thin-edge.io mode, if set to true authentication and registration will be done via thin-edge.
    enabled: true
    # MQTT client configuration, MQTT borker URL
    mqttServerURL: tcp://<<thinEdge MQTT broker>>
    # MQTT client configuration, client ID 

>     deviceId: My Gateway

Reviewer 2: - enabled is false by default, can you do the same here? 
- deviceId here should be the edge device name. Mqtt client ID desc. won't be understandable by the user and we need to mention that it should match with thinEdge device name. 
- There is a typo in mqttServerURL there.

Actually the application-default.yaml is more descriptive about all those fields and more understandable:
```
  thinEdge:
    # Enable thinEdge if OPC UA gateway is running next to thinEdge and should use thinEdge to connect to Cumulocity.
    # set enabled to false if OPC UA gateway is running without thinEdge
    enabled: false
    # MQTT Server URL of thinEdge (localhost)
    mqttServerURL: tcp://127.0.0.1:1883
    # Enables if mqtt client will use a steady and single connection. Beware that mqtt is only used to retrieve JWT, which is done depending how long JWT is valid. This is configurable (https://cumulocity.com/guides/device-sdk/mqtt/#jwt-token-retrieval)
    # It is recommended to use a steady connection only if the JWT is valid for a short time. Is the JWT for a longer time valid, standard 1h. It is not recommended to have a steady mqtt connection.
    mqttSteadyConnection: false
    # thinEdge deviceId must be changed, depends on the configured deviceId of the thin edge certificate, see https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/tutorials/connect-c8y.md#create-the-certificate
    deviceId: my-thin-edge-device
```

Author: Yes, you are right, i used the the documentation form application-default.yaml

      # automatically when the legacyCleanup is set to true, which is the default setting.
      # If the legacy files wanted to be kept or if the mechanism for clearing is not needed, set legacyCleanup to false.
      legacyCleanup: true

>   # These settings configure and enable/disable thin-edge mode (registration and operating OPC UA Gateway via thin-edge)

Reviewer: ```suggestion
  # These settings configure and enable/disable Thin Edge mode (registration and operating OPC UA Gateway via thin-edge)
```

Author: changed to your suggestions

>   # These settings configure and enable/disable Thin Edge mode (registration and operating OPC UA Gateway via thin-edge)

Reviewer: ```suggestion
  # These settings configure and enable/disable Thin Edge mode (registration and operating OPC UA gateway via Thin Edge).
```

  thinEdge:

>     # Enable thinEdge if OPC UA gateway is running next to thinEdge and should use thinEdge to connect to Cumulocity.

Reviewer: ```suggestion
    # Enable thinEdge if the OPC UA gateway is running next to Thin Edge and should use Thin Edge to connect to {{< product-c8y-iot >}}.
```

>     # set enabled to false if OPC UA gateway is running without thinEdge

Reviewer: ```suggestion
    # Set enabled to false if the OPC UA gateway is running without Thin Edge.
```

    enabled: false

>     # MQTT Server URL of thinEdge (localhost)

Reviewer: ```suggestion
    # MQTT Server URL of Thin Edge (localhost).
```

    mqttServerURL: tcp://127.0.0.1:1883

>     # Enables if mqtt client will use a steady and single connection. Beware that mqtt is only used to retrieve JWT, which is done depending how long JWT is valid. This is configurable (https://cumulocity.com/guides/device-sdk/mqtt/#jwt-token-retrieval)

Reviewer: ```suggestion
    # Enable this if the MQTT client uses a single steady connection. Note that MQTT is only used to retrieve the JWT, which is dependent on how long the JWT is valid. See https://cumulocity.com/guides/device-integration/mqtt/#jwt-token-retrieval.
```

We're in a code listing here, so we cannot use a normal reference:

`Refer to [Device integration using MQTT > Device certificates > JWT token retrieval](/device-integration/mqtt/#jwt-token-retrieval) in *Device integration* for information on how to configure this.`

i'm not sure if it is super helpful to add a URL that is not clickable.
how do we handle this in general? @BeateRixen

>     # It is recommended to use a steady connection only if the JWT is valid for a short time. Is the JWT for a longer time valid, standard 1h. It is not recommended to have a steady mqtt connection.

Reviewer: ```suggestion
    # We recommend you to use a steady connection only if the JWT is valid for a short time. If the JWT is valid for a longer time, the standard is one hour. It is generally not recommended to have a steady MQTT connection.
```

    mqttSteadyConnection: false

>     # thinEdge deviceId must be changed, depends on the configured deviceId of the thin edge certificate, see https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/tutorials/connect-c8y.md#create-the-certificate

Reviewer: ```suggestion
    # The Thin Edge deviceId must be changed, depending on the configured deviceId of the Thin Edge certificate, see https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/tutorials/connect-c8y.md#create-the-certificate.
```

see above, unsure if not clickable URLs are user-friendly.
